src/hotspot/share/gc/g1/g1CollectionSet.cpp

```diff
@@ -382,75 +378,60 @@ void G1CollectionSet::finalize_old_part(double time_remaining_ms) {
   QuickSort::sort(_collection_set_regions, _collection_set_cur_length, compare_region_idx);
 }
 
-void G1CollectionSet::move_candidates_to_collection_set(G1CollectionCandidateRegionList* regions) {
-  for (G1HeapRegion* r : *regions) {
-    _g1h->clear_region_attr(r);
-    add_old_region(r);
-  }
-  candidates()->remove(regions);
-}
-
 static void print_finish_message(const char* reason, bool from_marking) {
   log_debug(gc, ergo, cset)("Finish adding %s candidates to collection set (%s).",
                             from_marking ? "marking" : "retained", reason);
 }
 
-double G1CollectionSet::select_candidates_from_marking(double time_remaining_ms,
-                                                       G1CollectionCandidateRegionList* initial_old_regions,
-                                                       G1CollectionCandidateRegionList* pinned_old_regions) {
+double G1CollectionSet::select_candidates_from_marking(double time_remaining_ms) {
   uint num_expensive_regions = 0;
-
-  uint num_initial_regions_selected = 0;
-  uint num_optional_regions_selected = 0;
-  uint num_pinned_regions = 0;
+  uint num_inital_regions = 0;
+  uint num_initial_groups = 0;
+  uint num_optional_regions = 0;
 
   double predicted_initial_time_ms = 0.0;
   double predicted_optional_time_ms = 0.0;
 
   double optional_threshold_ms = time_remaining_ms * _policy->optional_prediction_fraction();
 
-  const uint min_old_cset_length = _policy->calc_min_old_cset_length(candidates()->last_marking_candidates_length());
-  const uint max_old_cset_length = MAX2(min_old_cset_length, _policy->calc_max_old_cset_length());
-  const uint max_optional_regions = max_old_cset_length - min_old_cset_length;
+  uint min_old_cset_length = _policy->calc_min_old_cset_length(candidates()->last_marking_candidates_length());
+  uint max_old_cset_length = MAX2(min_old_cset_length, _policy->calc_max_old_cset_length());
+  uint max_optional_regions = max_old_cset_length - min_old_cset_length;
   bool check_time_remaining = _policy->use_adaptive_young_list_length();
 
-  G1CollectionCandidateList* marking_list = &candidates()->marking_regions();
-  assert(marking_list != nullptr, "must be");
+  G1CSetCandidateGroupList* from_marking_groups = &candidates()->from_marking_groups();
 
   log_debug(gc, ergo, cset)("Start adding marking candidates to collection set. "
                             "Min %u regions, max %u regions, available %u regions"
                             "time remaining %1.2fms, optional threshold %1.2fms",
-                            min_old_cset_length, max_old_cset_length, marking_list->length(), time_remaining_ms, optional_threshold_ms);
+                            min_old_cset_length, max_old_cset_length, from_marking_groups->num_regions(), time_remaining_ms, optional_threshold_ms);
```

Comment:
Most of these debug messages are regions, it might be useful to add information about groups too. Otherwise it is impossible to understand any grouping issues.